### PR TITLE
3199 Separators can be added to context menus with this Handy New Trait

### PR DESCRIPTION
**Pull request by Cattlesquat**
image

Trait accepts:
(a) a text description
(b) a keystroke: if this keystroke is hidden by a Restrict Commands trait, then this separator will be hidden as well.

And that's it!

### CODE REVIEW

**Comment by Cattlesquat** [comment]
Oh and if you're "looking for it" while playing around with this, it's right next to "Sub Menu" in the Available Traits box.

**Comment by Cattlesquat** [comment]
Hi @yanlyub I need to know what this means I guess. It doesn't give me any hints where (or if) in my code this happens.

[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by net.sf.cglib.core.ReflectUtils$2 (file:/C:/Users/Brian%20Reynolds/.m2/repository/cglib/cglib-nodep/2.1_3/cglib-nodep-2.1_3.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
[ERROR] WARNING: Please consider reporting this to the maintainers of net.sf.cglib.core.ReflectUtils$2
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release

**Comment by Cattlesquat** [comment]
TIL that "git commit -a" doesn't pick up NEW files not previously in the build. And also that "GIANT ILLEGAL RED REFLEXIVE ACCESS OPERATION" is how that turns out ;-)   But hey, @yanlyub , your automated fuckery-detector did in fact get me to look & saved me a lot of WTF later when everybody says the PR doesn't work :D :D :D  Yay checks!!!

**Comment by uckelman** [comment]
> Hi @yanlyub I need to know what this means I guess. It doesn't give me any hints where (or if) in my code this happens.
> 
> [ERROR] WARNING: An illegal reflective access operation has occurred
> [ERROR] WARNING: Illegal reflective access by net.sf.cglib.core.ReflectUtils$2 (file:/C:/Users/Brian%20Reynolds/.m2/repository/cglib/cglib-nodep/2.1_3/cglib-nodep-2.1_3.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
> [ERROR] WARNING: Please consider reporting this to the maintainers of net.sf.cglib.core.ReflectUtils$2
> [ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
> [ERROR] WARNING: All illegal access operations will be denied in a future release
> 
> Everything was "apparently" building and running fine on my system, but then I got "errored" on the Travis check and so when I went back and did "mvn clean test" the above error (or "error warning", haha) was what I got.

It's not really an error. This comes from jmock-legacy needing to do something that Java presently considers "bad" in order to mock out a class in one of our tests. We might eventually get rid of it? Anyhow, it's not causing your problem.

**Comment by yanlyub** [comment]
And I thought the `-a` in `git commit -a` would add everything including new files.

I'm not such a git professional myself :)

And yes the error was something else, when you look for build errors, scroll all the way down in the build log (or click this round white "scroll to bottom" that hovers in the top right corner of the log), this particular build had compilation failures due to the missing file: https://travis-ci.com/github/vassalengine/vassal/builds/175608415#L8139

**Comment by Cattlesquat** [comment]
Yep is all fixed now, I just had to actually commit the "main file that this whole PR was all about" :D

**Comment by Cattlesquat** [comment]
@yanlyub You wanna know what adds everything including new files?

`git add *.*` is what :D

**Comment by uckelman** [comment]
I've already gotten rid of one of those illegal access "errors". I'm going to see if I can kill the other two.

**Comment by uckelman** [comment]
Those errors are gone in PR #92

File: vassal-app/src/main/java/VASSAL/counters/MenuSeparator.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, copies are available
+ * at http://www.opensource.org.
+ */
+
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/counters/MenuSeparator.java:38]: License notice goes at the top.

**Comment by uckelman** [comment]
Aside: The fact that a menu separator has to be in the trait stack and that requests for properties of traits below the menu separator have to go through it highlights how insane it is to use the decorator pattern for this...

**Comment by Cattlesquat** [comment]
Vassal has clearly "accreted" over a period of many eons :)

File: vassal-app/src/main/java/VASSAL/counters/MenuSeparator.java (hunk 1 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/MenuSeparator.java]: done

**Comment by Cattlesquat** [comment]
How long do you think it will take Vassal 4 to be "banged together from scratch"? (Both theoretical man hours and actual-calendar-time are relevant, though the latter is the real key). The veteran designer in me worries less about the actual barn-raising and more about all the "organic things" that will be unintentionally lost/discarded and have to be manually put back one by one (to get back to "at least as good as status quo")

**Comment by Cattlesquat** [comment]
I look forward to telling people "there's a slight perf hit for menu separators" :D

**Comment by uckelman** [comment]
I'll bet if you put about 30 menu separators at the top of one of your pieces, you'd be able to tell the difference.

**Comment by uckelman** [comment]
Ask me about V4 timelines again when I've had a clear week to pull together all my notes, page everything back into my head, and then present what I actually want to do.  I just can't say anything useful at the moment.

**Pull request merged**